core/src/main/java/org/elasticsearch/action/resync/TransportResyncReplicationAction.java

```diff
@@ -0,0 +1,158 @@
+        ResyncReplicationRequest request, IndexShard primary) throws Exception {
+        final ResyncReplicationRequest replicaRequest = performOnPrimary(request, primary);
+        return new WritePrimaryResult<>(replicaRequest, new ResyncReplicationResponse(), null, null, primary, logger);
+    }
+
+    public static ResyncReplicationRequest performOnPrimary(ResyncReplicationRequest request, IndexShard primary) {
+        return request;
+    }
+
+    @Override
+    protected WriteReplicaResult shardOperationOnReplica(ResyncReplicationRequest request, IndexShard replica) throws Exception {
+        Translog.Location location = performOnReplica(request, replica);
+        return new WriteReplicaResult(request, location, null, replica, logger);
+    }
+
+    public static Translog.Location performOnReplica(ResyncReplicationRequest request, IndexShard replica) throws Exception {
+        Translog.Location location = null;
+        for (Translog.Operation operation : request.getOperations()) {
+            try {
+                final Engine.Result operationResult = replica.applyTranslogOperation(operation, Engine.Operation.Origin.REPLICA,
+                    update -> {
+                        throw new TransportReplicationAction.RetryOnReplicaException(replica.shardId(),
+                            "Mappings are not available on the replica yet, triggered update: " + update);
+                    });
+                location = syncOperationResultOrThrow(operationResult, location);
+            } catch (Exception e) {
+                // if its not a failure to be ignored, let it bubble up
+                if (!TransportActions.isShardNotAvailableException(e)) {
+                    throw e;
+                }
+            }
+        }
+        return location;
+    }
+
+    @Override
+    public void sync(ResyncReplicationRequest request, Task parentTask, String primaryAllocationId,
+                     ActionListener<ResyncReplicationResponse> listener) {
+        // skip reroute phase
+        transportService.sendChildRequest(
+            clusterService.localNode(),
+            transportPrimaryAction,
+            new ConcreteShardRequest<>(request, primaryAllocationId),
+            parentTask,
+            transportOptions,
+            new TransportResponseHandler<ResyncReplicationResponse>() {
+                @Override
+                public ResyncReplicationResponse newInstance() {
+                    return newResponseInstance();
+                }
+
+                @Override
+                public String executor() {
+                    return ThreadPool.Names.SAME;
+                }
+
+                @Override
+                public void handleResponse(ResyncReplicationResponse response) {
+                    listener.onResponse(response);
+                }
+
+                @Override
+                public void handleException(TransportException exp) {
+                    listener.onFailure(exp);
```

Comment:
I handled it at top-level in IndexShard where the sync is triggered as primary failing / shutting down can also happen in the PrimaryReplicaSyncer. I think it's nicer not to rely on that though and I've added code here now to handle this.